helix-view/src/document.rs

```diff
@@ -163,23 +172,107 @@ impl Document {
         }
     }
 
+    #[inline(always)]
+    fn calculate_decode_capacity(decoder: &mut encoding_rs::Decoder, buf: &[u8]) -> usize {
+        // Will never panic because the buffer size will never be close to usize::MAX in the
+        // context it is used.
+        let valid_up_to = encoding_rs::Encoding::utf8_valid_up_to(buf);
+        let max_buffer_len = decoder
+            .max_utf8_buffer_length(buf.len() - valid_up_to)
+            .unwrap();
+        valid_up_to + max_buffer_len
+    }
+
+    #[inline(always)]
+    fn calculate_encode_capacity(encoder: &mut encoding_rs::Encoder, buf: &[u8]) -> usize {
+        // Will never panic because the buffer size will never be close to usize::MAX in the
+        // context it is used.
+        let valid_up_to = encoding_rs::Encoding::utf8_valid_up_to(buf);
+        let max_buffer_len = encoder
+            .max_buffer_length_from_utf8_if_no_unmappables(buf.len() - valid_up_to)
+            .unwrap();
+        (valid_up_to + max_buffer_len).next_power_of_two()
+    }
+
     // TODO: async fn?
-    pub fn load(path: PathBuf) -> Result<Self, Error> {
-        use std::{fs::File, io::BufReader};
+    /// Open a file from a path, or create a new file if it doesn't exist.
+    pub fn open(path: PathBuf) -> Result<Self, Error> {
```

Comment:
I think this function should have an optional encoding argument, which if not `None` forces it to attempt opening the file with the specified encoding.  It's not necessary right now, but presumably in the future we'll want a command for "attempt load/reload with specified encoding", so that the user can override when the auto-detect gets things wrong.  And an optional argument would pave the way for an easy implementation.